src/AWS.Deploy.CLI/Commands/DeployCommand.cs

```diff
@@ -128,32 +132,71 @@ public async Task<(Orchestrator, Recommendation, CloudApplication)> InitializeDe
 
             // Find existing application with the same CloudFormation stack name.
-            var deployedApplication = deployedApplications.FirstOrDefault(x => string.Equals(x.Name, cloudApplicationName));
+            var deployedApplication = allDeployedApplications.FirstOrDefault(x => string.Equals(x.Name, cloudApplicationName));
 
             Recommendation? selectedRecommendation = null;
             if (deployedApplication != null)
+            {
+                // Verify that the target application can be deployed using the current set of recommendations
+                if (!compatibleApplications.Any(app => app.StackName.Equals(deployedApplication.StackName, StringComparison.Ordinal)))
+                {
+                    var errorMessage = $"{deployedApplication.StackName} already exists as a Cloudformation stack but a compatible recommendation to perform a redeployment was no found";
+                    throw new FailedToFindCompatibleRecipeException(errorMessage);
+                }
+
                 // preset settings for deployment based on last deployment.
                 selectedRecommendation = await GetSelectedRecommendationFromPreviousDeployment(recommendations, deployedApplication, userDeploymentSettings);
+            } 
             else
-                selectedRecommendation = GetSelectedRecommendation(userDeploymentSettings, recommendations);
-
+            {
+                if (!string.IsNullOrEmpty(deploymentProjectPath))
+                {
+                    selectedRecommendation = recommendations.First();
+                }
+                else
+                {
+                    selectedRecommendation = GetSelectedRecommendation(userDeploymentSettings, recommendations);
+                }
+            }
+                
             var cloudApplication = new CloudApplication(cloudApplicationName, selectedRecommendation.Recipe.Id);
 
             return (orchestrator, selectedRecommendation, cloudApplication);
         }
 
+        /// <summary>
+        /// Filters the applications that can be re-deployed using the current set of available recommendations.
+        /// </summary>
+        /// <param name="allDeployedApplications"></param>
+        /// <param name="recommendations"></param>
+        /// <returns>A list of <see cref="CloudApplication"/> that are compatible for a re-deployment</returns>
+        private List<CloudApplication> GetCompatibleApplications(List<CloudApplication> allDeployedApplications, List<Recommendation> recommendations)
+        {
+            var compatibleApplications = new List<CloudApplication>();
+            foreach (var app in allDeployedApplications)
+            {
+                if (recommendations.Any(rec => string.Equals(rec.Recipe.Id, app.RecipeId, StringComparison.Ordinal)))
+                    compatibleApplications.Add(app);
+            }
+            return compatibleApplications;

```

Comment:
How does the tool ensure that deployed app is using a custom recipe?

A deployed app must use a unique recipe id which is saved along with `app.recipe` file during the first deployment.

In subsequent re-deploy, the generated deployment project must only be able to compatible with deployed app with that unique recipe id.

Are there test for it?